# Webshell Prediction/Detection

In this Notebook, we will train a classifier to identify malicious PHP files.

This is based on the discoveries in our other Notebooks that file entropy is strongly correlated with malicious obfuscated web shells, at entropy values unique from normal files in web roots across 3 common content management systems (Wordpress, Joomla, Drupal).

With this information, we will attempt predict malicious PHP based on entropy and a small selection of other features.

In [1]:
# Import a bunch of stuff
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.pipeline import make_pipeline
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler
import pandas as pd
import plotly.express as px
import numpy as np
import os

## Feature Creation

Since we're talking about files, there are no inherent features in our "dataset" as yet. We'll have to derive them one way or another. 

Other than entropy, some simple features we'll extract from the files include:

* Average line length
* Max line length
* Total line count

### PHP Only?

We could focus our dataset on just PHP files, but first I'd like to try to train against _all_ files that you might find in a CMS webroot. If that creates too many false positives, we'll limit to PHP.

## Data Import

We'll start by importing our data and configuring headers. We also have a helper function to extract file extensions for when we _do_ want to filter.

Our sample set is made of 3 entire CMS webroots along with 100+ samples of obfuscated PHP webshells.

In [3]:
header_names = [
    "filename",
    "path",
    "entropy",
    "is_elf",
    "md5",
    "sha1",
    "sha256",
    "sha512"
]

In [4]:
# Import individual CMS files and create the CMS category
drupal_df = pd.read_csv("drupal.csv", names=header_names)
drupal_df["cms"] = "drupal"
joomla_df = pd.read_csv("joomla.csv", names=header_names)
joomla_df["cms"] = "joomla"
wordpress_df = pd.read_csv("wordpress.csv", names=header_names)
wordpress_df["cms"] = "wordpress"
# Create main DF
df = pd.concat([drupal_df, joomla_df, wordpress_df]).reset_index()
df["disposition"] = "benign"

# Get webshells
webshell = pd.read_csv("webshell.csv", names=header_names)
webshell["disposition"] = "malicious"

# Add webshell to df
df = pd.concat([df, webshell]).reset_index()
df["extension"] = df.filename.apply(lambda f: os.path.splitext(f)[-1])

# Remove unnecessary cols
df.drop(["level_0","index","md5","sha1","sha256","sha512","is_elf","cms"], axis=1, inplace=True)

In [5]:
df

filename  \
0                                             .csslintrc   
1                                          .editorconfig   
2                                          .eslintignore   
3                                         .eslintrc.json   
4                                         .gitattributes   
...                                                  ...   
48455  r00tshell_6dc8b59781183d4061990b8b0fdb617063b8...   
48456  r57Shell_0963884cbc71293d7e290ad3ecf06a81355b5...   
48457   r57_89fa86a8748b5dfbabf610e47cf447675a817182.php   
48458  webshell_e8eaf8da94012e866e51547cd63bb99637969...   
48459  x2300Shell_e78c2c0ab5d6a9c6d381dfb51e4f38de159...   

                                                    path  entropy disposition  \
0                                  cms/drupal/.csslintrc     3.89      benign   
1                               cms/drupal/.editorconfig     4.68      benign   
2                               cms/drupal/.eslintignore     3.70      benign   
3                              cms/drupal/.eslintrc.json     4.14      benign   
4                              cms/drupal/.gitattributes     4.62      benign   
...                                                  ...      ...         ...   
48455  Webshell-samples/samples/webshell/PHP-backdoor...     5.71   malicious   
48456  Webshell-samples/samples/webshell/PHP-backdoor...     5.67   malicious   
48457  Webshell-samples/samples/webshell/PHP-backdoor...     6.02   malicious   
48458  Webshell-samples/samples/webshell/PHP-backdoor...     5.94   malicious   
48459  Webshell-samples/samples/webshell/PHP-backdoor...     5.67   malicious   

      extension  
0                
1                
2                
3         .json  
4                
...         ...  
48455      .php  
48456      .php  
48457      .php  
48458      .php  
48459      .php  

[48460 rows x 5 columns]

In [7]:
# Define feature functions
def avg_line_length(filepath: str) -> float:
    """
    Returns average line length for a file
    """
    with open(filepath) as f:
        try:
            return np.mean([len(l) for l in f.readlines()])
        except ValueError:
            return 1

def max_line_length(filepath: str) -> int:
    """
    Returns max line length for a file
    """
    with open(filepath) as f:
        try:
            return np.max([len(l) for l in f.readlines()])
        except:
            return None

    
def line_count(filepath: str) -> int:
    """
    Returns number of lines in a file
    """
    with open(filepath) as f:
        try:
            return len(f.readlines())
        except:
            return None

### Feature Generation

Now that we have the basics imported, it's time to generate our features by applying the functions defned above.

In [8]:
# Get features
# ============
# Get filesize
import os
df["size"] = df.path.apply(lambda f: os.path.getsize(f))
df["avg_line_length"] = df.path.apply(avg_line_length)
df["max_line_length"] = df.path.apply(max_line_length)
df["line_count"] = df.path.apply(line_count)
df["is_php"] = df.filename.str.endswith(".php")

/home/mttaggart/.cache/pypoetry/virtualenvs/webshell-entropy-XZZsjknK-py3.10/lib/python3.10/site-packages/numpy/core/fromnumeric.py:3464: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,


Now we clean the data by removing any features with `NaN` as a value.

In [9]:
clean = df.dropna().reset_index()

## Model Fit

With our source data ready, it's time to build and fit the model.

We begin by splitting the data into features (`X`) and the target `y`.

In [10]:
# Split the data
X = clean[["entropy","size","avg_line_length","max_line_length","line_count", "is_php"]]
y = clean.disposition

### Classification

Our initial model is very simple: a `KNeighborsClassifier` with a `StandardScaler` to normalize the features before fitting.

We'll further split the data betwen `train` and `test` sets.

In [12]:
# Instantiate our classified and scaler
knn = KNeighborsClassifier()
scaler = StandardScaler()

# Kfold
kf = KFold(n_splits=6, shuffle=True, random_state=42)

# Split the data between train/test
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, random_state=42)

# Make pipeline
pipeline = make_pipeline(scaler, knn)

Now we'll fit the model.

In [13]:
pipeline.fit(X_train, y_train)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('kneighborsclassifier', KNeighborsClassifier())])

With the model fit, now we make our predictions.

In [14]:
preds = pipeline.predict(X_test)

## Results Review

To fully review the results, we'll reattach the dropped columns we discarded for the model.

In [24]:
pred_df = X_test
pred_df["disposition"] = preds
pred_df["filename"] = [clean.loc[i, "filename"] for i in pred_df.index]
pred_df["path"] = [clean.loc[i, "path"] for i in pred_df.index]

Okay, what got flagged as malicious?

In [25]:
pd.set_option("display.max_colwidth", 300)
pred_df[pred_df.disposition == "malicious"]

entropy    size  avg_line_length  max_line_length  line_count  is_php  \
46622     6.11   17844        85.377990            766.0       209.0    True   
46598     5.96    1330        83.125000             90.0        16.0    True   
46560     5.71   38636        80.157676             92.0       482.0    True   
46617     6.07   23982       959.280000          23038.0        25.0    True   
46607     6.10    1389        81.705882             89.0        17.0    True   
46603     6.12    1281        80.062500             86.0        16.0    True   
46581     6.17    2006        80.240000             85.0        25.0    True   
46584     5.91    1718        85.900000             92.0        20.0    True   
46618     5.68  264001        81.886166             94.0      3224.0    True   
46639     5.75   20044        78.603922             92.0       255.0    True   
46609     6.10    1190        79.333333             87.0        15.0    True   
46549     6.03    8972      8972.000000           8972.0         1.0    True   
39133     5.13  274304       116.576286            205.0      2353.0    True   
46644     5.67  948460        80.968072             92.0     11714.0    True   
35737     4.55  548336        62.868150            156.0      8722.0    True   
46600     5.98    1180        84.285714             92.0        14.0    True   
46567     5.95     846       211.500000            747.0         4.0    True   
46643     5.94   49740        65.781746           6637.0       756.0    True   
46555     5.67  709804        81.954047             92.0      8661.0    True   
46587     6.17    1500        75.000000             85.0        20.0    True   
46586     5.88    2383       595.750000           2291.0         4.0    True   
46599     6.14    1247        77.937500             85.0        16.0    True   
46554     5.68  166907        81.817157             92.0      2040.0    True   
46636     5.67  654798        79.960679             92.0      8189.0    True   
46594     6.03    1394        82.000000             88.0        17.0    True   
46550     6.02   13774      6887.000000          13773.0         2.0    True   

      disposition                                                  filename  \
46622   malicious     WebShell_0ba8e8b6c1334b8335a9a9374bfb1109c0371478.php   
46598   malicious      Unknown_afdb2cf061897a383718ef8c59e9be10fc76d1c5.php   
46560   malicious      Sosyete_6ba719a5dcbbe675542d9001e5bf7987c979910f.php   
46617   malicious     WSOShell_1d2746a23a5201da7a0e89ff52adc3e8304b98a2.php   
46607   malicious      Unknown_d88733ed68ddb4a6f715d6a2c889d979c3ffef63.php   
46603   malicious      Unknown_c17433899e87dc065f3e6eadd26d92e82a62db08.php   
46581   malicious      Unknown_63e22f2dad6485eff1b516666443220c31a83173.php   
46584   malicious      Unknown_851cc138b197afae9e1f24d7ac8e7f7e89bff85d.php   
46618   malicious     WSOShell_606ece05d586d7b76817fbe10634871aa286222d.php   
46639   malicious              fabf134fce36292cf2bd03c5f2c9d3195f102bb7.php   
46609   malicious      Unknown_ead59aab3c6c886cf92b7e09b10d62c1db071eeb.php   
46549   malicious      DKShell_25d8e17abfa70370ab64ccc1f2e753a9082e8b7e.php   
39133   malicious                                     autoload_classmap.php   
46644   malicious   x2300Shell_e78c2c0ab5d6a9c6d381dfb51e4f38de159adf59.php   
35737   malicious                                                script.php   
46600   malicious      Unknown_b6f898e46c37a068ca52d3aa238be42ba86618fb.php   
46567   malicious      Unknown_0f06c5d1b32f4994c3b3abf8bb76d5468f105167.php   
46643   malicious     webshell_e8eaf8da94012e866e51547cd63bb996379690bf.php   
46555   malicious    GaZaShell_57dbe942519ec9b6a003e1cbcebb4dfe2c6c924d.php   
46587   malicious      Unknown_8ba4e5bbd60b384f855aa328a9a4f4e6c63007fb.php   
46586   malicious      Unknown_8af033424f9590a15472a23cc3236e68070b952e.php   
46599   malicious      Unknown_b5ba25ff0ec75761b0168f5c24f8419aa9e8903e.php   
46554   malicious  EkinoxShel

Just eyeballing it, all but 2 of our predictions are correct. That is pretty dang impressive!

And what was the score against our test set?

In [18]:
score = round(pipeline.score(X_test[["entropy","size","avg_line_length","max_line_length","line_count", "is_php"]], y_test) * 100, 2)
print(f"Score: {score}")
print(f"Cross val score:")
print(cross_val_score(pipeline, X, y, cv=kf))

Score: 99.97
Cross val score:
[0.99961415 0.99909956 0.99909956 0.99948546 0.99948546 0.9996141 ]


### Further detections

Now let's try with a completely new sample. This is a real webshell I've just encountered in the wild. It's been saved here as `webshell.php` Let's process its data into a DataFrame.

In [37]:
# Collect entropy!
entropy_result = ! sandfly-entropyscan/sandfly-entropyscan -file webshell.php 

In [39]:
# Build DataFrame
webshell_X = pd.DataFrame([{
    "entropy": float(entropy_result[2].split(": ")[-1]),
    "size": os.path.getsize("webshell.php"),
    "avg_line_length": avg_line_length("webshell.php"),
    "max_line_length": max_line_length("webshell.php"),
    "line_count": line_count("webshell.php"),
    "is_php": True
}])

And now, to predict.

In [41]:
# And our prediction...
pipeline.predict(webshell_X)

array(['malicious'], dtype=object)

## Conclusions

This is an extremely rudimentary model, but even with this approach, we can see that by adding some additional features easily derived from the files themselves, our detection rate goes up from 94.13% to 99.97%

Obfuscated PHP webshells are a fairly specific detection use case. However, the efficacy of this method speaks to the potential for well-applied machine learning in defense techniques.